#1312 ScalarArray code generation

mo-alistairp opened the pull request:
Adding code generation for ScalarArrays (towards #1312 ). The metadata support was added in #2173.

arporter commented on src/psyclone/domain/lfric/kern_call_acc_arg_list.py:
```diff
@@ -272,6 +272,20 @@ def scalar(self, scalar_arg, var_accesses=None):
+    def scalar_array(self, scalar_arr_arg, var_accesses=None):
+        '''
+        Override the default implementation as there's no need to specify
+        ScalarArrays for an OpenACC data region.
```
This isn't true I'm afraid - they will need to be added to a data region. However, you could leave this as a TODO as I'm not sure we're ever going to need data regions.

mo-alistairp replied:
Thanks for that - it's currently a very rough draft so some of the comments are currently just borrowed from other functions. But, I will definitely note that down!

arporter commented:
> I'm unsure how to add the declarations correctly. The ScalarArray needs to have `dimension(dims_array(1), dims_array(2), ... )`. I've tried creating each of `dims_array(n)` as individual symbols which get added to the ScalarArray symbols as a reference. However, this doesn't seem to be added to the dimension correctly and adds the `dims_array(n)` symbols to the reference list.

You'll need to create a Symbol for `dims_array` as an integer array that is an argument to the (kernel) subroutine. Then you create an ArrayType for the ScalarArray with `dims = [ArrayReference.create(dims_array_sim, [Literal(1)]), ArrayReference.create(dims_array_sym, [Literal(2)])]`.

mo-alistairp commented:
I think this is now ready for review. The code gen and stub gen seem to be performing as I would expect. I've added test coverage for everything I've added. Everything is passing the style checks.

arporter commented on src/psyclone/domain/lfric/lfric_scalar_array_args.py:
```diff
@@ -0,0 +1,313 @@
+
+    def _create_declarations(self):
+        '''
+        Add declarations for the ScalarArray arguments.
```
s/Add declarations/Create the Symbols/

mo-alistairp commented on src/psyclone/domain/lfric/kern_call_acc_arg_list.py:
```diff
@@ -271,6 +272,9 @@ def scalar(self, scalar_arg, var_accesses=None):
             :py:class:`psyclone.core.VariablesAccessMap`]
 
         '''
+        # TODO: Add implementation of OpenACC data region for ScalarArrays
```
Done

arporter commented on src/psyclone/domain/lfric/lfric_scalar_array_args.py:
```diff
@@ -0,0 +1,313 @@
+        super().__init__(node)
+
+        # Initialise dictionaries of 'real', 'integer' and 'logical'
+        # ScalarArray arguments by data type and intent
```
Thanks, please update the comment.

arporter commented on src/psyclone/domain/lfric/lfric_scalar_array_args.py:
```diff
@@ -181,93 +168,16 @@ def stub_declarations(self):
-            if self._logical_scalar_arrays[intent]:
-                for arg in self._logical_scalar_arrays[intent]:
+            for arg in self._scalar_array_args[intent]:
+                if arg.intent in FORTRAN_INTENT_NAMES:
```
Do you need this `if`? If so, please add a comment.

arporter commented on src/psyclone/tests/test_files/lfric/28.scalar_array_invoke.f90:
```diff
@@ -42,11 +42,11 @@ program scalar_array_invoke
 
```
Please update the comment in this file to give a bit more detail, i.e. that it exercises the ScalarArray functionality.

arporter commented on src/psyclone/domain/lfric/kern_call_arg_list.py:
```diff
@@ -252,6 +252,11 @@ def scalar(self, scalar_arg,
+            if scalar_arg.is_scalar_array:
+                # If it's a ScalarArray we need to add the dimensions
+                # array to the call
+                dims_sym = self._symtab.lookup("dims_" + scalar_arg.name)
```
Probably safer to use "dims_"+scalar_arg.name as a *tag* and then do `lookup_with_tag` here.

arporter commented:
> With regards to the change in `external/` - what is your process for getting something added there? I currently have a draft PR in `MetOffice/lfric_core` but I'm unsure if you need that to be committed first to get your integration tests to work

You can just update it as part of this PR. In order to get the integration tests to work, it will *also* need to be added to stfc/lfric_core. Ideally we'll remove this duplication and perhaps have a git submodule pointing to stfc/lfric_core but we're in a transition stage at the moment.

mo-alistairp commented:
> > With regards to the change in `external/` - what is your process for getting something added there? I currently have a draft PR in `MetOffice/lfric_core` but I'm unsure if you need that to be committed first to get your integration tests to work
> 
> You can just update it as part of this PR. In order to get the integration tests to work, it will _also_ need to be added to stfc/lfric_core. Ideally we'll remove this duplication and perhaps have a git submodule pointing to stfc/lfric_core but we're in a transition stage at the moment.

Ok, sure - thanks. Is it sufficient then if I wait until this is merged, then go through the merge process of the `MetOffice/lfric_core` PR so you can just update your stfc fork to main? Or does the stfc change need to be done sooner, e.g., by me opening another PR in stfc/lfric_core?
